lib/change_stream.js

```diff
@@ -0,0 +1,227 @@
+ * @function ChangeStream.prototype.next
+ * @param {ChangeStream~resultCallback} [callback] The result callback.
+ * @throws {MongoError}
+ * @return {Promise} returns Promise if no callback passed
+ */
+ChangeStream.prototype.next = function (callback) {
+  var self = this;
+  if (this.isClosed()) {
+    if (callback) return callback(new Error('Change Stream is not open.'), null);
+    return self.changeDomain.s.promiseLibrary.reject(new Error('Change Stream is not open.'));
+  }
+  if (typeof callback === 'function') {
+    return this.cursor.next(function(err, change) {
+      return processNewChange(self, err, change, callback);
+    });
+  }
+  return this.cursor.next().then(function(change) {
+    return processNewChange(self, null, change);
+  }).catch(function(err) {
+    return processNewChange(self, err, null);
+  });
+};
+
+/**
+ * Is the cursor closed
+ * @method ChangeStream.prototype.isClosed
+ * @return {boolean}
+ */
+ChangeStream.prototype.isClosed = function () {
+  if (!this.cursor) {
+    return true;
+  }
+  return this.cursor.isClosed();
+};
+
+/**
+ * Close the Change Stream
+ * @method ChangeStream.prototype.close
+ * @param {ChangeStream~resultCallback} [callback] The result callback.
+ * @return {Promise} returns Promise if no callback passed
+ */
+ChangeStream.prototype.close = function (callback) {
+  this.shuttingDown = true;
+  return this.cursor.close(callback);
+};
+
+// Handle new change events. This method brings together the routes from the callback, event emitter, and promise ways of using ChangeStream.
+var processNewChange = function (self, err, change, callback) {
+  // Handle errors
+  if (err) {
+    // Handle resumable MongoNetworkErrors if resumability is not disabled
+    if (!self.options.disableResume && err.name === 'MongoNetworkError') {
+      return self.cursor.close(function() {
+        if (typeof callback === 'function') return self.next(callback);
+        return self.next();
+      });
+    }
+
+    if (typeof callback === 'function') return callback(err, null);
+    if (typeof callback === 'string') return self.emit('error', err);
+    return self.changeDomain.s.promiseLibrary.reject(err);
+  }
+
+  // Cache the resume token if it is present. If it is not present and disableResume is false, return an error.
+  if (!self.options.disableResume) {
+    if (!change._id) {
+      var noResumeTokenError = new Error('A change stream document has been recieved that lacks a resume token (_id) and resumability has not been disabled for this change stream.');
+      if (typeof callback === 'function') return callback(noResumeTokenError, null);
+      if (typeof callback === 'string') return self.emit('error', noResumeTokenError);
+      return self.changeDomain.s.promiseLibrary.reject(noResumeTokenError);
+    }
+    self.cachedResumeToken = change._id;
+  }
+
+  // Close the change stream if an invalidation occurs
+  if (change.operationType === 'invalidate') {
+    self.close();
```

Comment:
the server will close the change stream in the event of an `invalidate` operation type